app/src/main/cpp/HandModels.cpp

```diff
@@ -0,0 +1,356 @@
+  return std::make_shared<vrb::ConcreteClass<HandModels, HandModels::State> >(aContext);
+}
+
+void
+HandModels::StartFrame() {
+}
+
+void
+HandModels::Draw(const vrb::Camera& aCamera, const Controller& aController) {
+  if (!aController.hasHandMesh)
+    return;
+
+  if (!m.program) {
+    VRB_ERROR("HandModel::Draw FAILED!");
+    return;
+  }
+
+  const HandGLState& state = m.handGLState.at(aController.index);
+
+  VRB_GL_CHECK(glUseProgram(m.program));
+
+  const GLboolean enabled = glIsEnabled(GL_DEPTH_TEST);
+  if (!enabled)
+    VRB_GL_CHECK(glEnable(GL_DEPTH_TEST));
+
+  glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+  glEnable(GL_BLEND);
+
+  VRB_GL_CHECK(glUniformMatrix4fv(m.uPerspective, 1, GL_FALSE, aCamera.GetPerspective().Data()));
+  VRB_GL_CHECK(glUniformMatrix4fv(m.uView, 1, GL_FALSE, aCamera.GetView().Data()));
+  // @FIXME: We don't apply any model transform for now, but eventually we could
+  // use `controller.transformMatrix`, if there is need.
+  vrb::Matrix modelMatrix = vrb::Matrix::Identity();
+  VRB_GL_CHECK(glUniformMatrix4fv(m.uModel, 1, GL_FALSE, modelMatrix.Data()));
+
+  assert(aController.meshJointTransforms.size() == state.jointCount);
+  std::vector<vrb::Matrix> jointMatrices;
+  jointMatrices.resize(state.jointCount);
+  for (int i = 0; i < state.jointCount; i++) {
+    jointMatrices[i] = aController.meshJointTransforms[i].PostMultiply(state.bindMatrices[i]);
+    VRB_GL_CHECK(glUniformMatrix4fv(m.uJointMatrices + i, 1, GL_FALSE, jointMatrices[i].Data()));
+  }
+
+  VRB_GL_CHECK(glBindBuffer(GL_ARRAY_BUFFER, state.vboPosition));
+  VRB_GL_CHECK(glEnableVertexAttribArray((GLuint)m.aPosition));
+  VRB_GL_CHECK(glVertexAttribPointer((GLuint)m.aPosition, 3, GL_FLOAT, GL_FALSE, 0, nullptr));
+
+  VRB_GL_CHECK(glBindBuffer(GL_ARRAY_BUFFER, state.vboNormal));
+  VRB_GL_CHECK(glEnableVertexAttribArray((GLuint)m.aNormal));
+  VRB_GL_CHECK(glVertexAttribPointer((GLuint)m.aNormal, 3, GL_FLOAT, GL_FALSE, 0, nullptr));
+
+  VRB_GL_CHECK(glBindBuffer(GL_ARRAY_BUFFER, state.vboJointIndices));
+  VRB_GL_CHECK(glEnableVertexAttribArray((GLuint)m.aJointIndices));
+  VRB_GL_CHECK(glVertexAttribPointer((GLuint)m.aJointIndices, 4, GL_FLOAT, GL_FALSE, 0, nullptr));
+
+  VRB_GL_CHECK(glBindBuffer(GL_ARRAY_BUFFER, state.vboJointWeights));
+  VRB_GL_CHECK(glEnableVertexAttribArray((GLuint)m.aJointWeights));
+  VRB_GL_CHECK(glVertexAttribPointer((GLuint)m.aJointWeights, 4, GL_FLOAT, GL_FALSE, 0, nullptr));
+
+  VRB_GL_CHECK(glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, state.iboIndices));
+  VRB_GL_CHECK(glDrawElements(GL_TRIANGLES, state.indexCount, GL_UNSIGNED_SHORT, nullptr));
+
+  glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
+  glBindBuffer(GL_ARRAY_BUFFER, 0);
+
+  if (!enabled)
+    VRB_GL_CHECK(glDisable(GL_DEPTH_TEST));
+}
+
+void HandModels::UpdateHandModel(const Controller& aController) {
+  VRB_LOG("HandModel: %u vertices, %u indices", aController.handMesh.vertexCount, aController.handMesh.indexCount);
```

Comment:
Not sure we need a logging statement here. In any case if you want to keep it, it should be more descriptive, perhaps adding the method name?

Reply:
That's a left-over, will remove it.